x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java

```diff
@@ -0,0 +1,239 @@
+import java.util.Base64;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.security.authc.esnative.tool.SetupPasswordTool.getErrorCause;
+
+public class PasswordAndEnrollmentInitialNode extends EnvironmentAwareCommand {
+    private static final Setting<SecureString> bootstrapPassword = SecureSetting.secureString("bootstrap.password",
+        null);
+    private static final Setting<SecureString> credentialsPassword = SecureSetting.secureString("bootstrap.password",
+        KeyStoreWrapper.SEED_SETTING);
+    private static final String elasticUser = ElasticUser.NAME;
+
+    private SecureString password;
+    private String token;
+    private String fingerprint;
+
+    // Protected for testing
+    protected SecureString getPassword() {
+        return password;
+    }
+
+    protected String getToken() {
+        return token;
+    }
+
+    protected String getFingerprint() {
+        return fingerprint;
+    }
+
+    protected CommandLineHttpClient getClient(Environment env) {
+        return new CommandLineHttpClient(env);
+    }
+
+    protected CreateEnrollmentToken getCreateEnrollmentToken(Environment env) throws Exception{
+        return new CreateEnrollmentToken(env);
+    }
+
+
+    PasswordAndEnrollmentInitialNode (){
+        super("Set elastic password and generate enrollment token for initial node");
+        parser.allowsUnrecognizedOptions();
+    }
+
+    public static void main(String[] args) throws Exception {
+        exit(new PasswordAndEnrollmentInitialNode().main(args, Terminal.DEFAULT));
+    }
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
+            throw new UserException(ExitCodes.NOOP, "This command is not intended for end users");
+        }
+        if (env.settings().hasValue(XPackSettings.ENROLLMENT_ENABLED.getKey()) && false ==
+            XPackSettings.ENROLLMENT_ENABLED.get(env.settings())) {
+            throw new UserException(ExitCodes.NOOP, "Enrollment is explicitly disabled.");
+        }
+        CommandLineHttpClient client = getClient(env);
+        CreateEnrollmentToken cet = getCreateEnrollmentToken(env);
+
+        // Try 5 times and rethrow the last exception from checkClusterHealth
+        for (int retry = 5; ; retry--) {
+            try {
+                checkClusterHealth(env, client);
+                break;
+            } catch (Exception e) {
+                if (retry == 1) {
+                    throw e;
+                }
+                Thread.sleep(1000);
+            }
+        }
+        if (Strings.isNullOrEmpty(bootstrapPassword.get(env.settings()).toString())) {
+            changeElasticUserPassword(env, terminal, client);
+        } else {
+            password = bootstrapPassword.get(env.settings());
+        }
```

Comment:
IMO this sort of high level design condition is unsuitable at this level of abstraction in the code. We are likely to introduce other checks like the one for the password hash set by the package installation script, and we should maybe reuse this cmd line utility for generating enrollment tokens apart from the first one.

I think this tool should receive by parameter whether it should reset the password in addition to generating the enrollment token (later we can add which types of enrollment tokens to generate).
If the arguments are incompatible with the configuration it should error out.

Lets discuss this sync.

Reply:
We will discuss it in our sync. Current version doesn't address it

Reply:
We decided this is indeed the good path.